gateway/ratelimiting/globalratelimiter.go

```diff
@@ -0,0 +1,140 @@
+package ratelimiting
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type globalRateState struct {
+	MaximumRequests uint64
+	Period          time.Duration
+
+	ResetTime time.Time
+	Requests  atomic.Uint64
+}
+
+type GlobalRateLimiter struct {
+	lock  sync.Mutex
+	state atomic.Pointer[globalRateState]
+}
+
+var _ RateLimiter = (*GlobalRateLimiter)(nil)
+
+func calculateAlignedResetTime(now time.Time, period time.Duration) time.Time {
+	// we align the start time to the beginning of the period
+	timeMs := now.UnixNano()
+	periodMs := int64(period / time.Nanosecond)
+	alignedNowMs := (timeMs / periodMs) * periodMs
+	alignedNow := time.Unix(0, alignedNowMs)
+	resetTime := alignedNow.Add(period)
+	return resetTime
+}
+
+func NewGlobalRateLimiter(maximumRequests uint64, period time.Duration) *GlobalRateLimiter {
+	// we align the start time to the beginning of the period
+	now := time.Now()
+	resetTime := calculateAlignedResetTime(now, period)
+
+	state := &globalRateState{
+		Period:          period,
+		MaximumRequests: maximumRequests,
+		ResetTime:       resetTime,
+	}
+
+	limiter := &GlobalRateLimiter{}
+	limiter.state.Store(state)
+
+	return limiter
+}
+
+func (l *GlobalRateLimiter) getState() *globalRateState {
+	state := l.state.Load()
+
+	now := time.Now()
+	if !now.Before(state.ResetTime) {
```

Comment:
nit: now.After(state.ResetTime) , if there is such a func, would be easier to read